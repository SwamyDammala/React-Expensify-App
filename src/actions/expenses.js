import uuid from 'uuid'
import database from '../firebase/firebase'

export const addexpense=(expense)=>({
       type:'Add_expense',
       expense
})

export const startAddExpense=(expenseData={})=>{
    return(dispatch)=>{
            //another way of defining default empty object entities
           const {
                description='',
                notes='',
                createdAt=0,
                amount=0
             }=expenseData
             const expense={description,amount,createdAt,notes}
                return   database.ref('expenses').push(expense).then((ref)=>{
                dispatch(addexpense({
                    id:ref.key,
                    ...expense
                }))
           })
    }
}

//Below is the remove the particular expense from expenses array based on the id. We are sending id to match and delete the item from expenses list
export const removeexpense=({id}={})=>({
    type:'Remove_expense',
    id
})

//Below is dunction to edit the exisiting expnase details in expenses array based on the expense id. We are sending update values as object and receive them as objects 
export const editexpense=(id,updates)=>({
    type:'Edit_expense',
    id,
    updates
})


//setExpenses
export const setExpenses=(expenses)=>({
    type:'SET_EXPENSES',
    expenses
})

export const setStartExpenses=()=>{
    return (dispatch)=>{
        return database.ref('expenses').once('value').then((snapshot)=>{
            const newExpenses=[]
            snapshot.forEach((childSnapshot)=>{
                newExpenses.push({
                    id:childSnapshot.key,
                    ...childSnapshot.val()
                })
            })
            dispatch(setExpenses(newExpenses))
        })
    }
}