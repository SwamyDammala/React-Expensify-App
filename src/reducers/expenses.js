const expenseReducerDefaultState=[]

//Below is the function we will pass into create store for expense array to update based on diferent types of conditions
const expenseReducer=(state=expenseReducerDefaultState,action)=>{
    switch(action.type){
                case 'Add_expense':            
                    return  [
                            ...state,
                            action.expense
                            ]
                case 'Edit_expense':
                        return state.map((expense)=>{
                            if(expense.id===action.id){
                                return{
                                    ...expense,
                                    ...action.updates
                                     } 
                                }
                            else{
                                return expense
                                }
                            })
                case 'Remove_expense':
                        return state.filter(({id})=>(action.id!==id))
                case 'SET_EXPENSES':
                    return  action.expenses
                    
                default:
                    return state            
            }
}

export default expenseReducer