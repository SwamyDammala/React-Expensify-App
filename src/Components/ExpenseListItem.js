import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import { removeexpense } from '../actions/expenses'
import  moment from 'moment'
import numeral from 'numeral'

const ExpenseListItem=({id,description,amount,createdAt,notes})=>(
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
            </Link>
            <p>
                {numeral(amount/100).format('$0,0.00')}
                -
                {moment(createdAt).format('MMMM Do, YYYY')}
                </p>
        {/* <button onClick={()=>{dispatch(removeexpense({id}))}}>Remove Expense</button> */}
    </div>
)




export default ExpenseListItem