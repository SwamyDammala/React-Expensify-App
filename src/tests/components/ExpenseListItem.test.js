import React from 'react'
import ExpenseListItem from '../../Components/ExpenseListItem'
import {shallow} from 'enzyme'
import expenses from '../fixtures/expenses'

const expense=expenses[0]

test('should test ExpensList Item for one of the object from expenses array',()=>{
    const wrapper=shallow(<ExpenseListItem {...expense} />)
    expect(wrapper).toMatchSnapshot()
})