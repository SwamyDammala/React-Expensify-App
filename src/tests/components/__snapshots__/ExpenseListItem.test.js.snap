// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test ExpensList Item for one of the object from expenses array 1`] = `
<div>
  <Link
    to="/edit/1"
  >
    <h3>
      gum
    </h3>
  </Link>
  <p>
    $4.00
    -
    January 1st, 1970
  </p>
</div>
`;
